src/klipse/plugin.cljs

```diff
@@ -105,24 +107,60 @@
   [element general-settings mode]
   (if-let [opts (@mode-options mode)]
     ;; weird piece of code - see klipsify-with-opts docstring
-    (go (<! ((<! (klipsify-with-opts element general-settings opts mode)))))
+    (go (<! ((<! (klipsify-with-opts element general-settings opts mode (chan (sliding-buffer 0)))))))
     (go (js/console.error "cannot find options for mode: " mode ". Supported modes: " (keys @mode-options)))))
 
-(defn ^:export klipsify-no-eval [element general-settings mode]
+(defn ^:export klipsify-no-eval [element general-settings mode channel]
   (if-let [opts (@mode-options mode)]
-    (klipsify-with-opts element general-settings opts mode)
+    (klipsify-with-opts element general-settings opts mode channel)
     (go #(go (js/console.error "cannot find options for mode: " mode ". Supported modes: " (keys @mode-options))))))
 
 (defn edit-elements [elements general-settings modes]
   (go
-    (loop [elements elements eval-fns []]
-      (if (seq elements)
-        (let [element (first elements)]
-          (if-let [mode (modes element)]
-            (let [eval-fn (<! (klipsify-no-eval element general-settings mode))]
-              (recur (rest elements) (conj eval-fns eval-fn)))
-            (recur (rest elements) eval-fns)))
-        eval-fns))))
+    (let [valid-elements (filter modes elements)
+          ; these channels are posted to whenever an edit is made
+          event-chans (mapv (fn [_] (chan 10)) valid-elements)
+          eval-fns (loop [elements valid-elements
+                          event-chans event-chans
+                          eval-fns []]
+                     (if (seq elements)
+                       (let [element (first elements)
+                             event-chan (first event-chans)]
+                         (let [eval-fn (<! (klipsify-no-eval element general-settings (modes element) event-chan))]
+                           (recur (rest elements) (rest event-chans) (conj eval-fns eval-fn))))
+                       eval-fns))
+          n (count eval-fns)
+          chan->idx (into {}
+                          (map vec (zip-colls event-chans (range))))]
+
+      ; subprocess that reloads other editors
+      (go
+        ; "edits" keeps track of the order in which user edits the editors
```

Comment:
why do you reload the snippets in the order in which user made edits to them?
I think it's simpler to reload all the snippets in the order they are in the page.

Reply:
The idea being that the order in which user edits the snippets dictates the dependency order, which might change as the user runs snippets multiple times.

Reply:
Interesting!
Could you show an example where it makes a difference and explain why your approach is preferable over the simpler approach of re-evaluating the snippets in the other they appear on the page?
It will help me decide what I want.

Reply:
* Open `clojure-reload-dbg.html?verbose=true`

* Change second snippet to `(def y 1)`

* Change first snippet to `(def x (inc y))`

* Change third snippet to `(def z (inc x))`, The second snippet will be reloaded before the first one (first depends on second)

* Change second snippet to `(def y 10)`, The snippets will be loaded in the order `1->0->2`, which is the correct according to the dependency order (`z->x->y`).

Simple approach has the benefit of being predictable. This approach infers the dependencies based on user interaction and does the correct thing as long as dependency order stays same. If it changes, it will take a couple of evals for algorithm to learn the new dependency order.

Reply:
I prefer to run the snippets in a simple order:
1. evaluate the current snippet
2. reevaluate all the snippets in the orders they appear on the page, including the current snippet.

What do you think?

Reply:
@viebel done